### PR TITLE
Improve coverage #24

### PR DESCRIPTION
**Pull request by asgrim**
Link to issue #24

### CODE REVIEW

**Comment by asgrim** [comment]
Note to @Ocramius: coverage is not complete with this PR, it will just help if we get progress so far merged in. We'll leave #24 open to continue to track coverage improvements

File: src/Identifier/IdentifierType.php [hunk 1]
```diff
@@ -46,11 +46,21 @@ public function getName()
      */
     public function getDisplayName()
     {
-        return ucfirst(basename($this->name));
+        return basename(str_replace('\\', '/', $this->name));
```
Comment (Ocramius) [inline src/Identifier/IdentifierType.php]: What is the reason for this change?

Comment (asgrim) [inline src/Identifier/IdentifierType.php]: because it didn't work before, writing the test discovered it did nothing ;)

File: src/Identifier/IdentifierType.php [hunk 2]
```diff
@@ -46,11 +46,21 @@ public function getName()
...
+    /**
+     * Check to see if a reflector is of a valid type specified by this identifier
+     *
+     * @param Reflection $reflector
+     * @return bool
+     */
     public function isMatchingReflector(Reflection $reflector)
     {
-        return $this->name === get_class($reflector);
+        if ($this->name == self::IDENTIFIER_CLASS) {
```
Comment (Ocramius) [inline src/Identifier/IdentifierType.php]: Confusing logic. What about other types of reflection?

File: src/Identifier/IdentifierType.php [hunk 3]
```diff
@@ -46,11 +46,21 @@ public function getName()
...
+            return $reflector instanceof ReflectionClass;
+        }
+
+        return false;
```
Comment (Ocramius) [inline src/Identifier/IdentifierType.php]: This code block can be collapsed into a single boolean expression

File: src/Identifier/IdentifierType.php (hunk 2 above)
Comment (asgrim) [inline src/Identifier/IdentifierType.php]: They will be added as and when they are supported (for example, in #40)

File: src/Identifier/IdentifierType.php (hunk 1 above)
Comment (Ocramius) [inline src/Identifier/IdentifierType.php]: So what is being done here? Is it just a user-friendly identifier?

File: src/Identifier/IdentifierType.php (hunk 3 above)
Comment (asgrim) [inline src/Identifier/IdentifierType.php]: It will be added to in #40 so no need for now

File: test/unit/Identifier/IdentifierTest.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+use BetterReflection\Identifier\Identifier;
+use BetterReflection\Identifier\IdentifierType;
+
+/**
+ * @covers \BetterReflection\Identifier\Identifier
+ */
+class IdentifierTest extends \PHPUnit_Framework_TestCase
+{
+    public function testValue()
```
Comment (Ocramius) [inline test/unit/Identifier/IdentifierTest.php]: Consider splitting into separate methods for the separate public APIs being used

File: test/unit/Identifier/IdentifierTypeTest.php [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+/**
+ * @covers \BetterReflection\Identifier\IdentifierType
+ */
+class IdentifierTypeTest extends \PHPUnit_Framework_TestCase
+{
+    public function possibleIdentifierTypesProvider()
+    {
+        return [
+            [IdentifierType::IDENTIFIER_CLASS, 'ReflectionClass'],
```
Comment (Ocramius) [inline test/unit/Identifier/IdentifierTypeTest.php]: `::class` please :)

File: src/Identifier/IdentifierType.php (hunk 1 above)
Comment (asgrim) [inline src/Identifier/IdentifierType.php]: yes, exactly, it turns `\BetterReflection\Reflection\ReflectionClass` into just `ReflectionClass` for example

File: test/unit/Identifier/IdentifierTypeTest.php [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+use BetterReflection\Identifier\IdentifierType;
+use BetterReflection\Reflection\ReflectionClass;
+use InvalidArgumentException;
+
+/**
+ * @covers \BetterReflection\Identifier\IdentifierType
+ */
+class IdentifierTypeTest extends \PHPUnit_Framework_TestCase
+{
+    public function possibleIdentifierTypesProvider()
```
Comment (Ocramius) [inline test/unit/Identifier/IdentifierTypeTest.php]: Docblock

File: test/unit/NodeCompiler/CompileNodeToValueTest.php [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+use BetterReflection\NodeCompiler\CompileNodeToValue;
+use LogicException;
+use PhpParser\Lexer;
+use PhpParser\Node\Expr\Yield_;
+use PhpParser\Parser;
+
+class CompileNodeToValueTest extends \PHPUnit_Framework_TestCase
+{
+    public function nodeProvider()
```
Comment (Ocramius) [inline test/unit/NodeCompiler/CompileNodeToValueTest.php]: Docblock

File: test/unit/Identifier/IdentifierTypeTest.php (hunk 5 above)
Comment (asgrim) [inline test/unit/Identifier/IdentifierTypeTest.php]: this is not the same as `\ReflectionClass`, so is not actually a class

File: src/Identifier/IdentifierType.php (hunk 2 above)
Comment (Ocramius) [inline src/Identifier/IdentifierType.php]: @asgrim add a `@TODO` then, please

File: test/unit/SourceLocator/StringSourceLocatorTest.php [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+        ));
+
+        $this->assertSame($sourceCode, $locatedSource->getSource());
+        $this->assertNull($locatedSource->getFileName());
+    }
+
+    public function testConstructorThrowsExceptionIfEmptyStringGiven()
+    {
+        $this->setExpectedException(InvalidArgumentException::class);
+        new StringSourceLocator('');
```
Comment (Ocramius) [inline test/unit/SourceLocator/StringSourceLocatorTest.php]: This seems to be invalid to me. Why is an empty string considered invalid source?

File: test/unit/SourceLocator/SingleFileSourceLocatorTest.php [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+        ));
+
+        $this->assertSame($expectedContent, $locatedSource->getSource());
+        $this->assertSame($fileName, $locatedSource->getFileName());
+    }
+
+    public function testConstructorThrowsExceptionIfEmptyFileGiven()
+    {
+        $this->setExpectedException(InvalidArgumentException::class);
+        new SingleFileSourceLocator('');
```
Comment (Ocramius) [inline test/unit/SourceLocator/SingleFileSourceLocatorTest.php]: More like "invalid file location"

File: test/unit/SourceLocator/LocatedSourceTest.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+        $this->assertSame($source, $locatedSource->getSource());
+        $this->assertSame($file, $locatedSource->getFileName());
+    }
+
+    public function testValuesWithNullFilename()
+    {
+        $source = '<?php echo "Hello world";';
+        $file = null;
+        $locatedSource = new LocatedSource($source, $file);
```
Comment (Ocramius) [inline test/unit/SourceLocator/LocatedSourceTest.php]: Is this for `eval()`'d code?

File: test/unit/SourceLocator/ComposerSourceLocatorTest.php [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            new IdentifierType(IdentifierType::IDENTIFIER_CLASS)
+        ));
+
+        $this->assertSame($expectedContent, $locatedSource->getSource());
+        $this->assertSame($fileName, $locatedSource->getFileName());
+    }
+
+    public function testInvokableThrowsExceptionWhenClassNotResolved()
+    {
+        $className = 'ClassWithNoNamespace';
```
Comment (Ocramius) [inline test/unit/SourceLocator/ComposerSourceLocatorTest.php]: `::class`?

File: src/Identifier/IdentifierType.php (hunk 1 above)
Comment (Ocramius) [inline src/Identifier/IdentifierType.php]: Maybe rename the method to make that clear?

File: src/Identifier/IdentifierType.php (hunk 3 above)
Comment (Ocramius) [inline src/Identifier/IdentifierType.php]: @asgrim add a `@TODO` then, please

File: test/unit/SourceLocator/ComposerSourceLocatorTest.php (hunk 11 above)
Comment (asgrim) [inline test/unit/SourceLocator/ComposerSourceLocatorTest.php]: Using `::class` here would load the class, which we do not want

File: test/unit/SourceLocator/StringSourceLocatorTest.php (hunk 8 above)
Comment (asgrim) [inline test/unit/SourceLocator/StringSourceLocatorTest.php]: There will be nothing in the AST if we load an empty string, so there is no point in even letting it past this initial point

File: test/unit/SourceLocator/LocatedSourceTest.php (hunk 10 above)
Comment (asgrim) [inline test/unit/SourceLocator/LocatedSourceTest.php]: Noooooo it does not get `eval()`'d, it gets loaded directly into the AST (rather than being loaded from a file)

File: test/unit/SourceLocator/StringSourceLocatorTest.php (hunk 8 above)
Comment (Ocramius) [inline test/unit/SourceLocator/StringSourceLocatorTest.php]: An empty string is still valid PHP source: it simply has no effect

File: test/unit/SourceLocator/LocatedSourceTest.php (hunk 10 above)
Comment (Ocramius) [inline test/unit/SourceLocator/LocatedSourceTest.php]: I'm confused by the meaning of `LocatedSource` then. Why is `LocatedSource('contents', null)` valid? What's the use-case scenario?

File: test/unit/SourceLocator/ComposerSourceLocatorTest.php (hunk 11 above)
Comment (Ocramius) [inline test/unit/SourceLocator/ComposerSourceLocatorTest.php]: Won't load, as per discussion. `::class` is a compile-time resolved expression that won't cause runtime side-effects

File: src/Identifier/IdentifierType.php (hunk 1 above)
Comment (asgrim) [inline src/Identifier/IdentifierType.php]: I've removed this method, probably not required anyway

File: test/unit/Identifier/IdentifierTest.php (hunk 4 above)
Comment (asgrim) [inline test/unit/Identifier/IdentifierTest.php]: Done

File: test/unit/Identifier/IdentifierTypeTest.php (hunk 6 above)
Comment (asgrim) [inline test/unit/Identifier/IdentifierTypeTest.php]: Done

File: test/unit/NodeCompiler/CompileNodeToValueTest.php (hunk 7 above)
Comment (asgrim) [inline test/unit/NodeCompiler/CompileNodeToValueTest.php]: Done

File: test/unit/SourceLocator/SingleFileSourceLocatorTest.php (hunk 9 above)
Comment (asgrim) [inline test/unit/SourceLocator/SingleFileSourceLocatorTest.php]: Added custom exception type

File: test/unit/SourceLocator/ComposerSourceLocatorTest.php (hunk 11 above)
Comment (asgrim) [inline test/unit/SourceLocator/ComposerSourceLocatorTest.php]: Fixed

File: test/unit/SourceLocator/LocatedSourceTest.php (hunk 10 above)
Comment (asgrim) [inline test/unit/SourceLocator/LocatedSourceTest.php]: For when you want to load source code directly from a string - the `StringSourceLocator` does exactly this, but the filename is allowed to be `null` (because you are loading from a string, rather than a file). It just means when you reflect from a class within a string, you'll not have a filename when you call `$classInfo->getFileName()` but then that would be expected (because you never loaded from file)

File: test/unit/SourceLocator/StringSourceLocatorTest.php (hunk 8 above)
Comment (asgrim) [inline test/unit/SourceLocator/StringSourceLocatorTest.php]: Added custom exception and comment to hopefully clarify

File: test/unit/SourceLocator/LocatedSourceTest.php (hunk 10 above)
Comment (Ocramius) [inline test/unit/SourceLocator/LocatedSourceTest.php]: @asgrim so why not using a  `StringSourceLocator` instead?

Comment (asgrim) [inline test/unit/SourceLocator/LocatedSourceTest.php]: Because this is a unit test, I'm just checking that the values are what we expect them to be when creating a `LocatedSource` VO

Comment (Ocramius) [inline test/unit/SourceLocator/LocatedSourceTest.php]: @asgrim still don't understand why `LocatedSource('content', null)` is valid: this means that the source was not "located", so the entire definition is borked IMO

Comment (asgrim) [inline test/unit/SourceLocator/LocatedSourceTest.php]: Created issue #43 and `@todo` comment

File: test/unit/SourceLocator/StringSourceLocatorTest.php (hunk 8 above)
Comment (Ocramius) [inline test/unit/SourceLocator/StringSourceLocatorTest.php]: Still think this invalid. Should I report a separate issue?

Comment (asgrim) [inline test/unit/SourceLocator/StringSourceLocatorTest.php]: It would not make any sense to accept an empty string as source. Otherwise, nothing else will stop you doing it, and it just won't find the thing you're trying to reflect, and you'll be like "WHY DOES MY THING NOT FIND REFLECTION" and 3 hours later you find it's because you accidentally passed an empty string to the `StringSourceLocator` because of a misnamed variable or something, and then you'll be like



However, I did forget to expect `EmptyPhpSourceCode` not `InvalidArgumentException` so I've pushed again.

Comment (Ocramius) [inline test/unit/SourceLocator/StringSourceLocatorTest.php]: Makes sense for now, but I will hit this use-case scenario when traversing a directory that contains empty files :-)
